File: backend/app/controllers/webhook_controller.py
from flask import Blueprint, request, jsonify
from services.webhook_service import handle_webhook_event
from services.auth_service import bind_line_uid_to_user_email, identify_google_user_by_token

webhook_bp = Blueprint("webhook", __name__)

@webhook_bp.route("/webhook", methods=["POST"])
def webhook():
    return handle_webhook_event(request.json)

@webhook_bp.route("/bind-google-email", methods=["POST"])
def bind_google_email():
    try:
        data = request.get_json()
        if not data:
            return jsonify({{"error": "no data"}}), 400
        for key in ('google_token', 'uid'):
            if key not in data:
                return jsonify({"error": f"缺少必要的資料({key})"}), 400
        line_uid = data.get("uid")
        user = identify_google_user_by_token(data)

        return bind_line_uid_to_user_email(line_uid, user)
    except Exception as e:
        return jsonify({"error": f"伺服器錯誤: {str(e)}"}), 500

@webhook_bp.route("/bind-email", methods=["POST"])
def bind_email():
    try:
        data = request.get_json()
        if not data:
            return jsonify({{"error": "no data"}}), 400
        for key in ('username', 'password', 'uid'):
            if key not in data:
                return jsonify({"error": f"缺少必要的資料({key})"}), 400
        line_uid = data.get("uid")
        user = identify_google_user_by_token(data)

        return bind_line_uid_to_user_email(line_uid, user)
    except Exception as e:
        return jsonify({"error": f"伺服器錯誤: {str(e)}"}), 500

File: backend/nest-app/src/controllers/webhook.controller.ts
import {
  Body,
  Controller,
  Post,
  Res,
  HttpStatus,
  Req,
} from '@nestjs/common';
import { Response } from 'express';
import { WebhookService } from '../services/webhook.service';
import { AuthService } from '../services/auth.service';
import { UserService } from '../services/user.service';
import { RedisService } from '../services/redis.service';
import { Request } from 'express';

@Controller()
  export class WebhookController {
    constructor(
      private readonly webhookService: WebhookService,
      private readonly authService: AuthService,
      private readonly userService: UserService,
      private readonly redisService: RedisService,
    ) {}

  @Post('webhook')
  async webhook(@Body() body: any, @Res() res: Response) {
    const timeout = new Promise((_, reject) =>
      setTimeout(() => reject(new Error('Request timeout')), 5000),
    );

    try {
      const result = await Promise.race([
        this.webhookService.handleWebhookEvent(body),
        timeout,
      ]);
      return res.json(result);
    } catch (error) {
      console.error('Webhook處理錯誤:', error);
      return res
        .status(HttpStatus.INTERNAL_SERVER_ERROR)
        .json({ error: `Webhook 處理發生錯誤: ${error.message || error.toString()}` });
    }
  }

  @Post('bind-google-email')
  async bindGoogleEmail(@Body() body: any, @Req() req: Request, @Res() res: Response) {
    if (!body) {
      return res.status(400).json({ error: 'no data' });
    }

    for (const key of ['google_token', 'uid']) {
      if (!body[key]) {
        return res.status(400).json({ error: `缺少必要的資料(${key})` });
      }
    }

    try {
      const user = await this.authService.identifyGoogleUserByToken(body);
      if (!user) {
        return res.status(400).json({ error: 'Google 用戶認證失敗' });
      }

      const bindResult = await this.authService.bindLineUidToUser(body.uid, user);
      const sessionId = this.userService.generateSignedSessionId(user.id);
      req.session.session_id = sessionId;
      await this.redisService.set(sessionId, user.id.toString(), 1800);

      return res
        .status(HttpStatus.OK)
        .cookie('session_id', sessionId, { httpOnly: true })
        .json(bindResult);
    } catch (e) {
      return res
        .status(500)
        .json({ error: `伺服器錯誤: ${e.message || e.toString()}` });
    }
  }

  @Post('bind-email')
  async bindEmail(@Body() body: any, @Req() req: Request, @Res() res: Response) {
    if (!body) {
      return res.status(400).json({ error: 'no data' });
    }

    for (const key of ['username', 'password', 'uid']) {
      if (!body[key]) {
        return res.status(400).json({ error: `缺少必要的資料(${key})` });
      }
    }

    try {
      const user = await this.authService.identifyGoogleUserByToken(body);
      if (!user) {
        return res.status(400).json({ error: 'Google 用戶認證失敗' });
      }

      const bindResult = await this.authService.bindLineUidToUser(body.uid, user);
      const sessionId = this.userService.generateSignedSessionId(user.id);
      req.session.session_id = sessionId;
      await this.redisService.set(sessionId, user.id.toString(), 1800);

      return res
        .status(HttpStatus.OK)
        .cookie('session_id', sessionId, { httpOnly: true })
        .json(bindResult);

    } catch (e) {
      return res
        .status(500)
        .json({ error: `伺服器錯誤: ${e.message || e.toString()}` });
    }
  }
}